packages/opencensus-core/src/metrics/gauges/gauge.ts

```diff
@@ -0,0 +1,170 @@
+   *
+   * It is recommended to keep a reference to the Point instead of always
+   * calling this method for manual operations.
+   *
+   * @param {LabelValue[]} labelValues The list of the label values.
+   * @returns {types.Point} The value of single gauge.
+   */
+  getOrCreateTimeSeries(labelValues: LabelValue[]): types.Point {
+    validateArrayElementsNotNull(
+        validateNotNull(labelValues, 'labelValues'), 'labelValue');
+    return this.registerTimeSeries(labelValues);
+  }
+
+  /**
+   * Returns a Point for a gauge with all labels not set, or default
+   * labels.
+   *
+   * @returns {types.Point} The value of single gauge.
+   */
+  getDefaultTimeSeries(): types.Point {
+    return this.registerTimeSeries(this.defaultLabelValues);
+  }
+
+  /**
+   * Removes the TimeSeries from the gauge metric, if it is present. i.e.
+   * references to previous Point objects are invalid (not part of the
+   * metric).
+   *
+   * @param {LabelValue[]} labelValues The list of label values.
+   */
+  removeTimeSeries(labelValues: LabelValue[]): void {
+    validateNotNull(labelValues, 'labelValues');
+    this.registeredPoints.delete(hashLabelValues(labelValues));
+  }
+
+  /**
+   * Removes all TimeSeries from the gauge metric. i.e. references to all
+   * previous Point objects are invalid (not part of the metric).
+   */
+  clear(): void {
+    this.registeredPoints.clear();
+  }
+
+  /**
+   * Registers a TimeSeries and returns a Point if the specified
+   * labelValues is not already associated with this gauge, else returns an
+   * existing Point.
+   *
+   * @param {LabelValue[]} labelValues The list of the label values.
+   * @returns {types.Point} The value of single gauge.
+   */
+  private registerTimeSeries(labelValues: LabelValue[]): types.Point {
+    const hash = hashLabelValues(labelValues);
+    // return if the specified labelValues is already associated with the point.
+    if (this.registeredPoints.has(hash)) {
+      return this.registeredPoints.get(hash);
+    }
+    if (this.labelKeysLength !== labelValues.length) {
+      throw new Error('Label Keys and Label Values don\'t have same size');
+    }
+
+    const point = new PointEntry(labelValues);
+    this.registeredPoints.set(hash, point);
+    return point;
+  }
+
+  /**
+   * Provides a Metric with one or more TimeSeries.
+   *
+   * @returns {Metric} The Metric.
+   */
+  getMetric(): Metric {
+    if (this.registeredPoints.size === 0) {
+      return null;
+    }
+    const hrtimeLocal = process.hrtime();
+    return {
+      descriptor: this.metricDescriptor,
+      timeseries: Array.from(
+          this.registeredPoints,
+          ([_, point]) => point.getTimeSeries(
+              {seconds: hrtimeLocal[0], nanos: hrtimeLocal[1]}))
+    };
+  }
+}
+
+/**
+ * The value of a single point in the Gauge.TimeSeries.
+ */
+export class PointEntry implements types.Point {
+  private value = 0;
+
+  constructor(private readonly labelValues: LabelValue[]) {
```

Comment:
my guess that `private readonly` should not be as argument

Reply:
@justindsmith any thoughts?

Reply:
Did a quick search, looks like we use a mix of both conventions throughout the codebase (fun!).

I'm not sure I would put a blanket "always" or "never" on this approach, but I think readability goes down when you have a mix of class-level and constructor-level properties defined that are both associated with the class. A developer now needs to look at _both_ places to see the complete set of properties. So I would say that if _all_ the properties of a class are passed in through the constructor we can use a visibility modifier to attach it to the class, but if we're defining other properties we shouldn't use that approach.

If we go by the above then in this case we should make it a class-level definition and not use the constructor-visiblity-modifier approach.

What do you all think of that? What are your thoughts on this?

Reply:
Just to make sure I understand your point.

> if all the properties of a class are passed in through the constructor we can use a visibility modifier to attach it to the class

something like this?
```
class PointEntry implements types.Point {
    private readonly labelValues: LabelValue[];

    constructor (labelValues: LabelValue[]) {
        this.labelValues = labelValues;
    }
}
```

Reply:
Yeah, in that case I think it's okay to just use the visibility modifier in the constructor since that value is the only class-level property.